google-cloud-firestore/src/main/java/com/google/cloud/firestore/UpdateBuilder.java

```diff
@@ -595,76 +594,98 @@ public T delete(@Nonnull DocumentReference documentReference) {
+    // 2. Build commit request
+    //
+    // Step 1 sets uses volatile property to ensure committed is visible to all
+    // threads immediately.
+    //
+    // Step 2 uses `forEach(..)` that is synchronized, therefore will be blocked
+    // until any writes are complete.
+    //
+    // Writes will verify `committed==false` within synchronized block of code
+    // before appending writes. Since committed is set to true before accessing
+    // writes, we are ensured that no more writes will be appended after commit
+    // accesses writes.
+    committed = true;
+    CommitRequest request = buildCommitRequest(transactionId);
+
     Tracing.getTracer()
         .getCurrentSpan()
         .addAnnotation(
             TraceUtil.SPAN_NAME_COMMIT,
-            ImmutableMap.of("numDocuments", AttributeValue.longAttributeValue(writes.size())));
-
-    final CommitRequest.Builder request = CommitRequest.newBuilder();
-    request.setDatabase(firestore.getDatabaseName());
-
-    for (WriteOperation writeOperation : writes) {
-      request.addWrites(writeOperation.write);
-    }
-
-    if (transactionId != null) {
-      request.setTransaction(transactionId);
-    }
-
-    committed = true;
+            ImmutableMap.of(
+                "numDocuments", AttributeValue.longAttributeValue(request.getWritesCount())));
 
     ApiFuture<CommitResponse> response =
-        firestore.sendRequest(request.build(), firestore.getClient().commitCallable());
+        firestore.sendRequest(request, firestore.getClient().commitCallable());
 
     return ApiFutures.transform(
         response,
         commitResponse -> {
-          List<com.google.firestore.v1.WriteResult> writeResults =
-              commitResponse.getWriteResultsList();
-
-          List<WriteResult> result = new ArrayList<>();
-
-          for (com.google.firestore.v1.WriteResult writeResult : writeResults) {
-            result.add(WriteResult.fromProto(writeResult, commitResponse.getCommitTime()));
-          }
-
-          return result;
+          Timestamp commitTime = commitResponse.getCommitTime();
+          return commitResponse.getWriteResultsList().stream()
+              .map(writeResult -> WriteResult.fromProto(writeResult, commitTime))
+              .collect(Collectors.toList());
         },
         MoreExecutors.directExecutor());
   }
 
+  private CommitRequest buildCommitRequest(ByteString transactionId) {
+    CommitRequest.Builder builder = CommitRequest.newBuilder();
+    builder.setDatabase(firestore.getDatabaseName());
+    forEach(builder::addWrites);
+    if (transactionId != null) {
+      builder.setTransaction(transactionId);
+    }
+    return builder.build();
+  }
+
   /** Checks whether any updates have been queued. */
   boolean isEmpty() {
-    return writes.isEmpty();
+    synchronized (writes) {
+      return writes.isEmpty();
+    }
   }
 
-  List<WriteOperation> getWrites() {
-    return writes;
+  void forEach(Consumer<Write> consumer) {
```

Comment:
nit: Consider renaming `forEach` to `forEachWrite` or something like that. Personally, I just find that `forEach` does not immediately convey _what_ it is iterating over.

Reply:
Done